codex/proof/datasampler.nim

```diff
@@ -0,0 +1,214 @@
+import std/bitops
+import std/sugar
+import std/sequtils
+
+import pkg/chronicles
+import pkg/chronos
+import pkg/questionable
+import pkg/questionable/results
+import pkg/constantine/math/arithmetic
+import pkg/poseidon2
+import pkg/poseidon2/io
+import pkg/libp2p
+import pkg/stew/arrayops
+
+import misc
+import slotblocks
+import types
+import datasamplerstarter
+import ../contracts/requests
+import ../blocktype as bt
+import ../merkletree
+import ../manifest
+import ../stores/blockstore
+import ../slots/converters
+import ../utils/digest
+
+# Index naming convention:
+# "<ContainerType><ElementType>Index" => The index of an ElementType within a ContainerType.
+# Some examples:
+# SlotBlockIndex => The index of a Block within a Slot.
+# DatasetBlockIndex => The index of a Block within a Dataset.
+
+logScope:
+  topics = "codex datasampler"
+
+type
+  DataSampler* = ref object of RootObj
+    slot: Slot
+    blockStore: BlockStore
+    slotBlocks: SlotBlocks
+    # The following data is invariant over time for a given slot:
+    datasetRoot: Poseidon2Hash
+    slotRootHash: Poseidon2Hash
+    slotTreeCid: Cid
+    datasetToSlotProof: Poseidon2Proof
+    blockSize: uint64
+    numberOfCellsInSlot: uint64
+    datasetSlotIndex: uint64
+    numberOfCellsPerBlock: uint64
+
+proc getNumberOfCellsInSlot*(slot: Slot): uint64 =
+  (slot.request.ask.slotSize.truncate(uint64) div DefaultCellSize.uint64)
+
+proc new*(
+    T: type DataSampler,
+    slot: Slot,
+    blockStore: BlockStore
+): DataSampler =
+  # Create a DataSampler for a slot.
+  # A DataSampler can create the input required for the proving circuit.
+  DataSampler(
+    slot: slot,
+    blockStore: blockStore
+  )
+
+proc start*(self: DataSampler): Future[?!void] {.async.} =
+  let slotBlocks = SlotBlocks.new(
+    self.slot.slotIndex,
+    self.slot.request.content.cid,
+    self.blockStore)
+  if err =? (await slotBlocks.start()).errorOption:
+    error "Failed to create SlotBlocks object for slot", error = err.msg
+    return failure(err)
+
+  let
+    manifest = slotBlocks.manifest
+    numberOfCellsInSlot = getNumberOfCellsInSlot(self.slot)
+    blockSize = manifest.blockSize.uint64
+
+  without starter =? (await startDataSampler(self.blockStore, manifest, self.slot)), e:
```

Comment:
you can initialize the builder with a verifiable manifest and it will unpack and set the slots and root hashes.